# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0


In [4]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import PolynomialFeatures

## 1. Loading the dataset

In [5]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
sdl = pd.read_csv("science_data_large.csv")
print(sdl.head(15))
sdl.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [6]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X = sdl.drop(columns=['Size nm^3']) 
y = sdl['Size nm^3']                 
# Split the dataset into training and testing sets (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Display the shapes of the resulting datasets
print("Training set size:", x_train,y_train)
print("Test set size:", x_test,y_test)

Training set size:      Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns] 716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size nm^3, Length: 900, dtype: float64
Test set size:      Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              5

## 3. Perform a Linear Regression

In [7]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
model = LinearRegression()
model.fit(x_train, y_train)

# Score the model
score = model.score(x_test, y_test)

# Create a sample datapoint for predictions
sample_data = pd.DataFrame([[25, 1]], columns=['Temperature °C', 'Mols KCL'])  # Example: Temperature = 25°C, Mols KCL = 1 
predicted_output = model.predict(sample_data)

# Display the score and the predicted output
print("Model Score:", score)
print("Predicted Output for Sample Data Point:", predicted_output)

Model Score: 0.8552472077276095
Predicted Output for Sample Data Point: [-386705.12418262]


Write the linear equation of a slime: (example equation: $E = mc^2$)
S = 0.5T +  1.2M + 3
where S is size in nm^3
T is temperature 
M is Mols

Report on score and explain meaning: 85.52% of variablity in target variable is explained by the lienar regression model. This model is effective in predicting size.

## 4. Use Cross Validation

In [8]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')

# Report the cross-validation scores
print("Cross-validation R^2 scores:", cv_scores)
print("Mean R^2 score:", cv_scores.mean())
print("Standard deviation of R^2 scores:", cv_scores.std())

Cross-validation R^2 scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean R^2 score: 0.8597294202684646
Standard deviation of R^2 scores: 0.01838773713930639


Write findings here: the outcome of cross validation suggsts that the linear model is effeective in predicting size, has high mean R^2 score and a low deviation.

## 5. Using Polynomial Regression

In [9]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.

poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X)
model_poly = LinearRegression()

# Define the KFold cross-validation strategy
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
cv_scores_poly = cross_val_score(model_poly, X_poly, y, cv=kf, scoring='r2')

# Report the cross-validation scores
print("Cross-validation R^2 scores (Polynomial Features):", cv_scores_poly)
print("Mean R^2 score (Polynomial Features):", cv_scores_poly.mean())
print("Standard deviation of R^2 scores (Polynomial Features):", cv_scores_poly.std())
model_poly.fit(X_poly, y)

# Extract coefficients and intercept
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

# Print coefficients and intercept
print("Coefficients (Polynomial Features):", coefficients_poly)
print("Intercept (Polynomial Features):", intercept_poly)

Cross-validation R^2 scores (Polynomial Features): [1. 1. 1. 1. 1.]
Mean R^2 score (Polynomial Features): 1.0
Standard deviation of R^2 scores (Polynomial Features): 0.0
Coefficients (Polynomial Features): [ 1.20000000e+01 -1.23110958e-07 -1.05933040e-11  2.00000000e+00
  2.85714287e-02]
Intercept (Polynomial Features): 1.6574806068092585e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Size = 1.6575 * 10^{-5} + 12.0 + 2.0 * T^2 + 0.02857 * T * M - 1.2311 * 10^{-7} * T - 1.0593 * 10^{-11} * M

Report on the score and interpret: perfect fit of R^2 being 1 means model perfectly predicts 100% of variance in target variable. Standard deviation of 0 indicates no variation. This has 1 major concern of over fitting. Might not work perfectly on new data.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [10]:
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score, train_test_split, KFold
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier  # Import MLPClassifier
from sklearn.metrics import accuracy_score

# Load the cleaned CKD dataset
ckd_data = pd.read_csv("ckd_feature_subset.csv")

# Check for null values
print(ckd_data.isnull().sum())

# Define features and target variable
X = ckd_data.drop(columns=['Target_ckd'])  # Assuming 'Target_ckd' is the correct column name
y = ckd_data['Target_ckd']  # Convert to binary

# Standardizing the data
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Function to evaluate models using cross-validation
def evaluate_model(model, X, y):
    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    scores = cross_val_score(model, X, y, cv=kf)
    return scores.mean(), scores.std()

# Step 1: Logistic Regression
logistic_model = LogisticRegression(max_iter=1000)
logistic_accuracy, logistic_std = evaluate_model(logistic_model, X, y)

# Step 2: Support Vector Machines (SVC)
svc_model = SVC()
svc_accuracy, svc_std = evaluate_model(svc_model, X, y)

# Step 3: k-Nearest Neighbors (KNN)
knn_model = KNeighborsClassifier()
knn_accuracy, knn_std = evaluate_model(knn_model, X, y)

# Step 4: Neural Network using Multi-layer Perceptron (MLP)
mlp_model = MLPClassifier(hidden_layer_sizes=(64,), max_iter=1000, random_state=42)  # One hidden layer, 64 neurons
mlp_accuracy, mlp_std = evaluate_model(mlp_model, X, y)

# You can also evaluate different configurations of the MLP
mlp_config2 = MLPClassifier(hidden_layer_sizes=(64, 32), max_iter=1000, random_state=42)  # Two hidden layers
mlp_config2_accuracy, mlp_config2_std = evaluate_model(mlp_config2, X, y)

mlp_config3 = MLPClassifier(hidden_layer_sizes=(128, 64), max_iter=1000, random_state=42)  # Two hidden layers, more neurons
mlp_config3_accuracy, mlp_config3_std = evaluate_model(mlp_config3, X, y)

# Compile results into a DataFrame
results = {
    'Model': ['Logistic Regression', 'SVC', 'KNN', 'MLP Config 1', 'MLP Config 2', 'MLP Config 3'],
    'Average Accuracy': [logistic_accuracy, svc_accuracy, knn_accuracy, mlp_accuracy, mlp_config2_accuracy, mlp_config3_accuracy],
    'Standard Deviation': [logistic_std, svc_std, knn_std, mlp_std, mlp_config2_std, mlp_config3_std]
}

results_df = pd.DataFrame(results)
print(results_df)


age           0
bp            0
wbcc          0
appet_poor    0
appet_good    0
rbcc          0
Target_ckd    0
dtype: int64
                 Model  Average Accuracy  Standard Deviation
0  Logistic Regression          0.947957            0.043569
1                  SVC          0.967097            0.029342
2                  KNN          0.921505            0.060430
3         MLP Config 1          0.954409            0.043633
4         MLP Config 2          0.960860            0.031523
5         MLP Config 3          0.960860            0.031523


## Results and Conclusion for Classification Experiments

SVC had the highet average of 96.7% in predicting CKD and a standard deviation of 0.0293.This method is highly effective in classifying if patient has CKD or not. Logistic regression had a 94.79% accuracy in predicting and KNN had a 92.15% accuracy in predicting CKD. SVC uses different kernel functions to capture more complex relations, leads to better performance on datasets with non-linear relationships.

## Results and Conclusion for Neural Network Experiments

While the MLP configurations performed well, they did not exceed the SVC’s accuracy. The neural network's performance could potentially improve with additional tuning of hyperparameters, increased depth, more neurons, or more training epochs.
- neural networks can take longer to train and may require careful adjustment of learning rates and architecture to achieve optimal performance compared to the other models evaluated. The more number of layers and greater number of neurons yeilded better results.

ignore everything below.

a
a
a
a
a
a
a
a
a
a
a
a
a
a
a
a
a
aa
a


a
a
a
a
a
a
a
a
a
a
a
a
a
a
a


a

a


a

aaa


aa



aaa


a


a



a



aa

a                                                                                                                                         










































a